sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala

```diff
@@ -180,6 +229,55 @@ case class Join(
     }
   }
 
+  override protected def validConstraints: Set[Expression] = {
+    // Currently we only propagate constraints if the condition consists of equality
+    // and ranges. For all other cases, we return an empty set of constraints
+    def constructIsNotNullConstraints(condition: Expression): Set[Expression] = {
```

Comment:
We typically avoid nested methods unless you are actually closing over state from the nested scope in the function.  I'd consider just making this a private method.